mmocr/models/textrecog/layers/transformer_layer.py

```diff
@@ -283,12 +282,143 @@ def forward(self, x):
         return x
 
 
+class PositionAttention(nn.Module):
+    """Transcribes visual features into character probabilities. A part of
+    implementation of `ABINet <https://arxiv.org/abs/1910.04396>`_. Adapted
+    from https://github.com/FangShancheng/ABINet .
+
+    Args:
+        max_length (int): Maximum length :math:`T` of generated text sequence.
+        in_channels (int): Number of channels :math:`E` of input vector.
+        num_channels (int): Number of channels of hidden vectors in mini U-Net.
+        h (int): Height :math:`H` of input image features.
+        w (int): Width :math:`W` of input image features.
+        mode (str): Upsampling mode for :obj:`torch.nn.Upsample` in mini U-Net.
+            Defaults to "nearest".
+    """
+
+    def __init__(self,
+                 max_length,
+                 in_channels=512,
+                 num_channels=64,
+                 h=8,
+                 w=32,
+                 mode='nearest',
+                 **kwargs):
+        super().__init__()
+        self.max_length = max_length
+
+        self.k_encoder = nn.Sequential(
+            self._encoder_layer(in_channels, num_channels, stride=(1, 2)),
+            self._encoder_layer(num_channels, num_channels, stride=(2, 2)),
+            self._encoder_layer(num_channels, num_channels, stride=(2, 2)),
+            self._encoder_layer(num_channels, num_channels, stride=(2, 2)))
+
+        self.k_decoder = nn.Sequential(
+            self._decoder_layer(
+                num_channels, num_channels, scale_factor=2, mode=mode),
+            self._decoder_layer(
+                num_channels, num_channels, scale_factor=2, mode=mode),
+            self._decoder_layer(
+                num_channels, num_channels, scale_factor=2, mode=mode),
+            self._decoder_layer(
+                num_channels, in_channels, size=(h, w), mode=mode))
```

Comment:
This part can be abstract as miniUnet, which is for the domain transfer for key
```suggestion
        self.k = miniUnet(                 
                 in_channels=3,
                 base_channels=num_channels,
                 num_stages=4,
                 strides=((1, 2), (2, 2), (2, 2), (2, 2), (2, 2)),
                 scale_factor=(2,2,2,None),
                 size=(None, None, None, (h, w)),
                 mode=(mode, mode, mode, mode)
                 )
```
reference the implement in mmocr'unet

Reply:
The structure is very simple and adding it to backbones might be unnecessary. We can leave it as it is until some other tasks have used a similar structure.

Reply:
It does not need to add it to backbones. Actually, it is just a nonlinear transform from feature to key. But written all in the class, it's difficult to understand it immediately， while abstracting it as `self.k_project = miniUnet()`, it's no need to read miniUnet for knowing this function. Also, essentially, PositionAttention is a standard Attention, except q is PositionalEncoding, which should be passed to attention, and key is projected with a complex net.

Reply:
I still don't think we should add mini-unet.
1. In `forward()`, there are sufficient comments and clearly divided code blocks for users to understand what these layers are doing.
2. I think we are mostly sure that it won't be reused by other network implementations. 
3. The suggested interface is overly complicated and increases the cost of understanding for users. Imagine that a user has to jump between files to understand how this network has been created, then they realize that it can be clearly presented within 8 lines. 

Overall, I think the **reusability** matters the most - if we don't believe that this sub-network can significantly benefit other models and therefore it is worth being more generalized, we can save efforts on separating it out from the current module. Instead, we just make sure comments are enough to keep users from being lost.

Reply:
The decode used transformer in MMOCR,  q = nonlinear(feat)(position), k = nonlinear(feat)(complex | optional) v = feat and DotProductAttentionLayer(q, k, v mask)

Reply:
if it won't be reused by other network implementations, it may merge the PositionAttention into ABIVisionDecoder directly, not building a PositionAttention class individually